### ISSUE TITLE
RELAYMSGs appear like external messages and due to lack of ZWSP cause self-pings

### ISSUE THREAD

**Issue by Mikaela** [hifi/heisenbridge#153]

**Comment by poVoq** [comment]
You are referring to external RELAYMSG, right? Sadly Heisenbridge itself doe not do RELAYMSG in the botbridge mode yet AFAIK.

**Comment by hifi** [comment]
Yes, I believe so. RELAYMSGs are sent by users "not in the channel" from the perspective of Heisenbridge so it shows them as external messages via notice from the heisenbridge bot so it will definitely cause all kinds of pings.

The real fix is to add generic relaybot support into Heisenbridge where it will have a configurable delay when ghosts created by RELAYMSG source nicks and other parsed relaybot sources.

**Comment by Mikaela** [comment]
Yes, I am referring to external RELAYMSGs, e.g.:

> `Message from mikaela-feneas-org/m!~u@b5g7r52kdqb7i.irc: As PrivacyTools.io is moving, nothing written today should say via=privacytools.io to avoid easily avoidable dead homeserver in the future. Maybe once the new homeserver is known if it will appear in those rooms`

would to a IRC client appear as a normal PRIVMSG from nickname `mikaela-feneas-org/m` and thus not cause a ping.

**Comment by poVoq** [comment]
Any news on this? Would be really nice if Heisenbridge would work better in plumbmode, both handling external RELAYMSG better and maybe hopefully do RELAYMSG itself at some point.

**Comment by hifi** [comment]
Sending RELAYMSGs is rather trivial so I could start with that. One thing that needs to be done is a conversion from MXID/displayname to IRC safe nickname for RELAYMSG.

Receiving is a bit more hairy as there are two modes: relay and ghosts.

If using ghosts there needs to be a timer to clean them up at some point if the relayed user does not say anything for a long time.

If I first add a simple relay mode support for received RELAYMSGs (any nicks containing a forward slash) then it would wook quite a bit better to begin with and the messages just appear as coming from Heisenbridge like so:

```
Heisenbridge: <user/network> message
```

**Comment by poVoq** [comment]
It's mainly about other chat-systems bridged via Matterbridge utilizing RELAYMSG.

Your suggestion would be for sure an improvement over the apparently broken state right now, but at least for me personally the added benefit of running Heisenbridge over just keeping Matterbridge also posting to a Matrix channel (also <user/network> like) would be too minimal to be worth it.

**Comment by hifi** [comment]
We'll get to proper ghosts eventually as just I need to iterate in small increments. Adding metadata (are they relayed and last time they spoke) to ghosts is the biggest obstacle but certainly doable.

I pushed the first version of generic CAP support earlier as the first step to sending RELAYMSGs.

**Comment by hifi** [comment]
#189 implements sending and would appreciate testing it. I haven't found any wild implementations of the draft other than Ergo so that's what I've tested so far.

**Comment by poVoq** [comment]
Have tested it with Conduit. Sending via RELAYMSG works fine now, but it still seems to conflict with other services such as Matterbridge that can also use RELAYMSG.
